[CI:BUILD] Packit: add jobs for downstream Fedora package builds

lsm5 opened the pull request:
Get rid of `skopeo.spec.rpkg` in favour of
`rpm/skopeo.spec` which gets synced with fedora dist-git on every upstream release. The version in the new spec file is set to `0` by default and gets updated by packit automatically on every packit task.

Packit will automatically create a PR on fedora dist-git on every new upstream release. A sample PR will look like:
https://src.fedoraproject.org/rpms/container-selinux/pull-request/10#

A dry run for this can be triggered using:
`$ packit propose-downstream --local-content`

To run this command locally, you would need to have your packit user-configuration-file set.
Ref: https://packit.dev/docs/configuration/#user-configuration-file

along with a fedora api key created at:
https://src.fedoraproject.org/settings#nav-api-tab with sufficient ACLs.

This patch incurs zero additional cost to upstream as the downstream
tasks occur only after upstream release and do not block anything
upstream.


[NO NEW TESTS NEEDED]

lsm5 commented:
@mtrmac @vrothberg @rhatdan @edsantiago @cevich PTAL

lsm5 commented:
Successful copr build at: https://copr.fedorainfracloud.org/coprs/rhcontainerbot/packit-builds/build/6058022/

cevich commented on rpm/update-spec-provides.sh:
```diff
@@ -0,0 +1,17 @@
+
+sed -i '/Provides: bundled(golang.*/d' $SPEC_FILE
+
+GO_IMPORTS=$(golist --imported --package-path github.com/containers/$PACKAGE --skip-self | sort -u | xargs -I{} echo "Provides: bundled(golang({}))")
```
I'm not sure in this case, but might confirm that `-I{}` doesn't need to be quoted to protect it from bash gobbling it up.

cevich commented on rpm/skopeo.spec:
```diff
@@ -0,0 +1,194 @@
+%changelog
+%if %{with manual_changelog}
+* Mon May 01 2023 RH Container Bot <rhcontainerbot@fedoraproject.org>
+- Placeholder changelog for envs that are not autochangelog-ready
```
Since it's a placeholder it's probably minor, but is `RH Container Bot <rhcontainerbot@fedoraproject.org>` still relevant here (since automation is going to packit)?  Might consider updating this (and the date) if not.

lsm5 replied:
Changelog entry was copy-pasta from podman spec. This changelog entry is only so that copr builds on centos 8 stream do not complain about no changelog. The copr repos are still owned by `rhcontainerbot` user. 

The changelog entry is not expected to be updated regularly, and might even go away once centos 8 stream gets autochangelog support.

cevich commented on rpm/skopeo.spec:
```diff
@@ -0,0 +1,194 @@
+export CGO_CFLAGS=$CFLAGS
+
+%if %{without btrfs}
+export BUILDTAGS="exclude_graphdriver_btrfs btrfs_noversion"
```
In `.cirrus.yml` the BUILDTAGS are `'btrfs_noversion libdm_no_deferred_remove containers_image_openpgp`.  Is the value in the spec file correct?

lsm5 replied:
hmm, this is what we've had in the fedora spec file all this while. Let me get back on this.

lsm5 replied:
I guess I can just include what the upstream makefile includes by default. Looking in the current [skopeo.spec.rpkg](https://github.com/containers/skopeo/blob/main/skopeo.spec.rpkg#L96) on main, it only has libdm_tag.sh apart from the btrfs tags.

lsm5 replied:
I can try to use Makefile targets for rpm builds as well and see how those go.

lsm5 replied:
giving up on Makefile targets in %build section for now. Adjusted buildtags to match upstream Makefile's.

cevich reviewed (approved):
LGTM - other than a few total-guesses at things that might need fixing.

edsantiago commented on rpm/skopeo.spec:
```diff
@@ -0,0 +1,194 @@
+
+%prep
+%autosetup -Sgit %{name}-%{version}
+sed -i 's/install-binary: bin\/%{name}/install-binary:/' Makefile
+sed -i 's/completions: bin\/%{name}/completions:/' Makefile
+sed -i 's/install-docs: docs/install-docs:/' Makefile
```
These squick me out a little. Best I can tell, they're removing dependencies. Can you document why? And maybe use `.*` instead of knowing the rest of the command line? And maybe also anchor with `^`?

lsm5 replied:
fixed in upcoming..

lsm5 commented on rpm/skopeo.spec:
```diff
@@ -0,0 +1,194 @@
+
+%install
+make \
+    PREFIX=%{buildroot}%{_prefix} \
```
PREFIX in makefile is set to /usr/local, so that's why setting `PREFIX=%{buildroot}%{_prefix}` takes of DESTDIR too, but anyway, changed this to `DESTDIR=%{buildroot} PREFIX=%{_prefix}` in upcoming..

mtrmac replied:
Yes, right now this does not matter, because the spec file does not run the top-level `install` target.

(That target writes into `$destdir/etc`, so overwriting PREFIX would not work.)

At this point it’s just a matter of precisely matching the semantics to avoid the risk of future surprises.

lsm5 commented on rpm/skopeo.spec:
```diff
@@ -0,0 +1,194 @@
+%endif
+
+export GOPATH=$(pwd)/_build:$(pwd)
+export CGO_CFLAGS=$CFLAGS
```
great catch! fixed in upcoming..

lsm5 commented on rpm/skopeo.spec:
```diff
@@ -0,0 +1,194 @@
+
+# go-rpm-macros package and autochangelog exists for fedora and rhel9
+%if 0%{?fedora} || 0%{?rhel} >= 9
+%bcond_without go_rpm_macros
```
these should be read in the opposite sense actually. See: https://rpm-software-management.github.io/rpm/manual/conditionalbuilds.html  . If you think that's gonna keep confusing people long term, I could probably switch to simply `%bcond` or some other conditional and get rid of the `with and without`. I added a note in the spec along with the link.

mtrmac replied:
My mistake, thanks for the pointer.

I think this counts as “expected RPM knowledge”, no need to make the actual spec file mechanism more complicated for ignorant readers like me.

lsm5 commented:
I'm gonna try disabling `.packit.sh` and see how things go. Please hold off further reviews.

mtrmac commented on .packit.sh:
```diff
@@ -5,22 +5,28 @@
+git archive --prefix=$PACKAGE-$RPM_VERSION/ -o $PACKAGE-$VERSION.tar.gz HEAD
+
+# rpmbuild expects source tarball in the same dir as spec file
+ln -s $PACKAGE-$VERSION.tar.gz rpm/
```
```suggestion
ln -s ../$PACKAGE-$VERSION.tar.gz rpm/
```
I think

lsm5 replied:
this file has been deleted.

mtrmac commented on rpm/skopeo.spec:
```diff
@@ -0,0 +1,206 @@
+export CGO_CFLAGS="$CGO_CFLAGS -m64 -mtune=generic -fcf-protection=full"
+%endif
+
+export BASEBUILDTAGS="$(hack/libdm_tag.sh) $(hack/libsubid_tag.sh)"
```
(Non-blocking: If the variable is only consumed immediately below, `export` seems unnecessary.)

lsm5 replied:
thanks! updated.

lsm5 commented on rpm/skopeo.spec:
```diff
@@ -0,0 +1,194 @@
+ln -s ../../../../ src/%{import_path}
+cd ..
+
+mv vendor src
```
This is needed for the non-copr official koji builds where the build envs don't have network access.

Without this, I see a build failure like: https://kojipkgs.fedoraproject.org//work/tasks/369/102110369/build.log

mtrmac replied:
My immediate thought is to blame it on the `GO111MODULE=off`; `-mod=vendor` is documented to be the default and to not hit the network.

But I didn't actually test that for a long time, I might get to experiment next week.

lsm5 replied:
The `%gobuild` macro provided by Fedora sets `GO111MODULE=off`.

I guess I could try setting %gobuild in the spec file for all environments, but I'm afraid we would miss out on any changes made by the toolchain team.

mtrmac replied:
https://pagure.io/go-rpm-macros/issue/28 suggests just overriding that.


Intuitively going back to managing `GOPATH` manually seems like losing progress, but I really don’t have the full context to appreciate which of the two approaches is, overall, more workable or maintainable. I’ll fully leave that up to you.

lsm5 replied:
let me give that a quick try

lsm5 replied:
@mtrmac that worked!! Thanks for bothering to look that up :) . I had stuck around with it for too long to care. Spec file is much simpler now. Other comments addressed as well.

lsm5 commented on rpm/skopeo.spec:
```diff
@@ -0,0 +1,194 @@
+export CGO_CFLAGS="$CGO_CFLAGS -m64 -mtune=generic -fcf-protection=full"
+%endif
+
+export GOPATH=$(pwd)/_build:$(pwd)
```
looks like it still is. See comment: https://github.com/containers/skopeo/pull/2011#discussion_r1228609843

lsm5 commented on .packit.sh:
```diff
@@ -5,22 +5,28 @@
+git archive --prefix=$PACKAGE-$RPM_VERSION/ -o $PACKAGE-$VERSION.tar.gz HEAD
+
+# rpmbuild expects source tarball in the same dir as spec file
+mv $PACKAGE-$VERSION.tar.gz rpm/
```
I'm so happy you made this comment. I see that with the spec file switch: `skopeo.spec.rpkg`  -> `rpm/skopeo.spec` we can just use packit's defaults, so we don't need this script. Got rid of it in the latest version of the PR. Thanks a lot!

lsm5 commented:
> alright, I think I have addressed all comments. Added some additional comments in the spec file and also got rid of `.packit.sh`. Packit's defaults are good enough.
> 
> PTAL.

err, once again I spoke too soon. https://copr.fedorainfracloud.org/coprs/rhcontainerbot/packit-builds/package/skopeo/ something seems wrong with version fetching.

lsm5 commented:
> err, once again I spoke too soon. https://copr.fedorainfracloud.org/coprs/rhcontainerbot/packit-builds/package/skopeo/ something seems wrong with version fetching.

So, using packit's defaults means the Version string is fetched using: `git describe --tags --abbrev=0` and not using the string in `version/version.go`. So the current build has version `1.12.0` and not `1.12.1~dev`.  If that's acceptable, then we should be ok. I will upgrade the `Epoch:102` so existing copr users are not broken.

I guess upstream itself doesn't need to concern itself too much with all this, more of a copr usability and breakage thing.

mtrmac commented on rpm/skopeo.spec:
```diff
@@ -0,0 +1,209 @@
+# DO NOT TOUCH the Version string!
+# The TRUE source of this specfile is:
+# https://github.com/containers/skopeo/blob/main/rpm/skopeo.spec
+# If that's what you're reading, Version must be 0, and will be updated by Packit for
```
The value is not `0` right now, maybe the comment needs updating.

lsm5 replied:
argh, I'll fix that. thanks

lsm5 commented:
> I will upgrade the `Epoch:102` so existing copr users are not broken.

set `Epoch: 102` to preserve upgrade path for copr users. With that I think we should be good to go.

mtrmac commented on rpm/skopeo.spec:
```diff
@@ -0,0 +1,209 @@
+# unset LDFLAGS earlier set from set_build_flags
+LDFLAGS=''
+
+# These steps are needed for Koji builds
```
This does not quite help explain the rationale. Admittedly the PR discussion document that now; still, if we have a comment, it seems worth to make that comment useful stand-alone.

Maybe something like
> We want to use the Fedora-provided %%gobuild, which sets `GO111MODULE=off; so we need to set up a
> GOPATH environment to use the vendored dependencies instead of trying to download them from the
> network (which is prohibited by Koji)

?

lsm5 replied:
SGTM, will update. Thanks

mtrmac reviewed (commented):
> So the current build has version 1.12.0 and not 1.12.1~dev. If that's acceptable, then we should be ok.

The currently-generated version numbers use the snapshot syntax, that’s a clear enough indication of what the sources are.

lsm5 commented:
@mtrmac all comments should be addressed now. Marking ready for review. PTAL.

mtrmac reviewed (approved):
LGTM. Thanks!
